106-auto-rebalance-on-imbalanced-clusters.md

```diff
@@ -0,0 +1,471 @@
+    RebalanceOnScaleDown-->>RebalanceOnScaleUp: then scale up
+    RebalanceOnScaleUp-->>RebalanceOnAnomalyDetection: then skew rebalance
+  end
+  RebalanceOnScaleUp->>Idle: scale up complete with no queued items
+```
+
+* from **RebalanceOnScaleUp**:
+  * **RebalanceOnScaleUp**: if a rebalancing on scale up is still running or another one was requested while the first one ended.
+  * **RebalanceOnScaleDown**: if a scale down operation was requested, so the current rebalancing scale up is stopped (and queued) and a new rebalancing scale down is started. The rebalancing scale up will be postponed.
+  * **RebalanceOnAnomalyDetection**: if a configmap related to goal violation was detected. It will run once the queued scale down and scale up is completed.
+  * **Idle**: if a scale up operation was requested, it was executed and completed successfully or failed.
+
+```mermaid
+sequenceDiagram
+  autonumber
+  RebalanceOnAnomalyDetection->>RebalanceOnScaleDown: queued scale down
+  RebalanceOnAnomalyDetection->>RebalanceOnScaleUp: queued scale up
+  RebalanceOnAnomalyDetection->>RebalanceOnAnomalyDetection: queued skew rebalance
+  critical queued scale up and skew rebalance
+    RebalanceOnAnomalyDetection->>RebalanceOnScaleUp: execute scale up
+    RebalanceOnScaleUp-->>RebalanceOnAnomalyDetection: then skew rebalance
+  end
+  critical queued scale down and skew rebalance
+    RebalanceOnAnomalyDetection->>RebalanceOnScaleDown: execute scale down
+    RebalanceOnScaleDown-->>RebalanceOnAnomalyDetection: then skew rebalance
+  end
+  critical queued scale down and scale up and skew rebalance
+    RebalanceOnAnomalyDetection->>RebalanceOnScaleDown: execute scale down
+    RebalanceOnScaleDown-->>RebalanceOnScaleUp: then scale up
+    RebalanceOnScaleUp-->>RebalanceOnAnomalyDetection: then skew rebalance
+  end
+  RebalanceOnAnomalyDetection->>Idle: scale up complete with no queued items
+```
+* from **RebalanceOnAnomalyDetection**:
+  * **RebalanceOnAnomalyDetection**: if another goal violation was detected while the first one ended. If a scale down and scale up is also queued up then they will execute first.
+  * **RebalanceOnScaleUp**: if a rebalancing on scale up is queued and will run if there is no other rebalancing scale down in queue. If a rebalancing scale down is in queue then it will be executed first.
+  * **RebalanceOnScaleDown**: if a scale down operation was requested, it will run once the skew rebalance is completed
+  * **Idle**: if skew rebalance was requested, it was executed and completed successfully or failed.
+
+On each reconciliation, the following process will be used:
+
+```mermaid
+flowchart TB
+  A[KafkaClusterCreator] --creates--> B[KafkaCluster]
+  B -- calls --> D[KafkaAutoRebalancingReconciler.reconcile]
+  D -- check for configmap with goal-violation prefix --> E{if config map present?}
+  D -- if rebalance in progress --> F[ignore new configmaps and delete them]
```

Comment:
which CMs will be deleted in that case? My understanding is that `KafkaAutoRebalancingReconciler` will not create new ones. What about these CMs that are used for current rebalancing? These should be deleted by rebalancing itself or?